### PR TITLE
[#171211343] User data processing api

### PR DESCRIPTION
**Pull request by AleDore**
Fixed model for user data processing api.
This PR is necessary to proceed with other dependent Pull requests in io-functions-app.
Once merged wit master it could be useful to publish a new version of nom package in order to align all others project inter-dependencies

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>Please include a Pivotal story at the beginning of the PR title (see below).</td>
    </tr>
  </tbody>
</table>



  Example of PR titles that include pivotal stories:

  * single story: `[#123456] my PR title`
  * multiple stories: `[#123456,#123457,#123458] my PR title`

    
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 59c7b8da2e745c868b1ccf3c53e22877fe9c2ddb
</p>

**Comment by gunzip** [comment]
it looks like linting does not pass on circleci see https://app.circleci.com/jobs/github/pagopa/io-functions-commons/935 can you fix this @AleDore ?

**Comment by AleDore** [comment]
fix lint on circleCi done!

File: openapi/definitions.yaml [hunk 1]
```diff
@@ -637,6 +637,33 @@ ServiceTuple:
...
+    status:
+      $ref: "#/UserDataProcessingStatus"
+    createdAt:
+      $ref: "#/Timestamp"
+    updatedAt:
+      $ref: "#/Timestamp"
+    version:
+      type: integer
+  required:
+    - id
```
Comment (gunzip) [inline openapi/definitions.yaml]: since id is auto generated server side it cannot be required here

File: src/models/user_data_processing.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+export const RetrievedUserDataProcessing = tag<IRetrievedUserDataProcessing>()(
+  t.intersection([
+    UserDataProcessing,
+    DocumentDbUtils.RetrievedDocument,
+    VersionedModel
+  ])
+);
+
+// tslint:disable-next-line: prettier
```
Comment (gunzip) [inline src/models/user_data_processing.ts]: why do we need this ?

File: src/models/user_data_processing.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    userDataProcessing: UserDataProcessing
+  ): Promise<Either<QueryError, UserDataProcessing>> {
+    const newId = makeUserDataProcessingId(
+      userDataProcessing.fiscalCode,
+      userDataProcessing.choice
+    );
+
+    const toUpdate: UserDataProcessing = {
+      userDataProcessingId: newId,
+      // tslint:disable-next-line: object-literal-sort-keys
```
Comment (gunzip) [inline src/models/user_data_processing.ts]: can we just sort the keys here ? =)

File: openapi/definitions.yaml [hunk 4]
```diff
@@ -637,6 +637,33 @@ ServiceTuple:
...
+    id:
+      type: string
+      minLength: 1
+    fiscalCode:
+      $ref: "#/FiscalCode"
+    choice:
+      $ref: "#/UserDataProcessingChoice"
+    status:
+      $ref: "#/UserDataProcessingStatus"
+    createdAt:
```
Comment (gunzip) [inline openapi/definitions.yaml]: properties must be snake_cased into openapi specs (created_at, updated_at, fiscal_code, ...)

File: src/models/user_data_processing.ts (hunk 3 above)
Comment (AleDore) [reply src/models/user_data_processing.ts]: Could be fine this way? -> 
`public async createOrUpdateByNewOne(
    userDataProcessing: UserDataProcessing
  ): Promise<Either<QueryError, UserDataProcessing>> {
    const newId = makeUserDataProcessingId(
      userDataProcessing.choice,
      userDataProcessing.fiscalCode
    );

    const toUpdate: UserDataProcessing = {
      choice: userDataProcessing.choice,
      createdAt: userDataProcessing.createdAt,
      fiscalCode: userDataProcessing.fiscalCode,
      status: userDataProcessing.status,
      updatedAt: userDataProcessing.createdAt,
      userDataProcessingId: newId
    };`

**Comment by gunzip** [comment]
it looks like test fail: https://app.circleci.com/jobs/github/pagopa/io-functions-commons/964

File: openapi/definitions.yaml [hunk 5]
```diff
@@ -637,6 +637,29 @@ ServiceTuple:
...
+      $ref: "#/Timestamp"
+    updated_at:
+      $ref: "#/Timestamp"
+    version:
+      type: integer
+  required:
+    - fiscal_code
+    - choice
+    - status
+    - created_at
```
Comment (gunzip) [inline openapi/definitions.yaml]: it looks like the API caller has to provide a create_at value to insert a new request. is it possible to make this field computed server side ? (same for update_at)

Comment (AleDore) [reply openapi/definitions.yaml]: Yes it is possible but I' don' t know if it is mandatory to mantain the created_date generation on client side (app) for legal purpose.

**Comment by gunzip** [comment]
it isn't

"I'm not anti-social; I'm just not user friendly"


Il giorno lun 24 feb 2020 alle ore 12:06 AleDore <notifications@github.com>
ha scritto:

File: openapi/definitions.yaml [hunk 6]
```diff
@@ -674,3 +696,30 @@ HttpsUrl:
...
+    The status of a user data processing request
+  enum:
+    - PENDING
+    - WIP
+    - CLOSED
+  example: PENDING
+UserDataProcessingChoice:
+  type: string
+  description: |-
+    User's choice to delete or download its own data.
```
Comment (gunzip) [inline openapi/definitions.yaml]:
```suggestion
    User's choice to delete or download his own data.
```

**Pull request merged**